### PR TITLE
Add more data to output of `collStats` and `dbStats`

### LINKED ISSUE #2447: Add more  data to the output of `collStats` and `dbStats` commands
### What should be done?

Capped collection related stats can be tested after https://github.com/FerretDB/FerretDB/issues/3458.

Currently, we return a limited number of fields for `collStats` requests. We should support more fields, see https://www.mongodb.com/docs/manual/reference/command/collStats/.

See TODO in the code.

Add 
- `collstats` `capped`
- `collstats` `max`
- `collstats` `maxSize`
- `collStats` `indexDetails`
- `collStats` `indexBuilds`
- `collStats` `inMemory`
- `collStats` `numOrphanDocs`
- `aggregate` `$collstats` `capped`
- `aggregate` `$collstats` `indexDetails`
- `aggregate` `$collstats` `indexBuilds`
- `dbStats` `sizeIndexFreeStorage`
- `dbStats` `fsUsedSize`
- `dbStats` `fsTotalSize`
- `dbStats` `operationTime`

Out of scope
- https://github.com/FerretDB/FerretDB/issues/3582

### LINKED ISSUE #3521: Fix failing stats tests for the new PostgreSQL backend
### What should be done?

Figure out why tests fail and provide the necessary fixes.

### Where?

```
task test-integration-postgresql SHARD_INDEX=0 SHARD_TOTAL=0 TEST_RUN='TestCommandsAdministrationCollStats'
task test-integration-postgresql SHARD_INDEX=0 SHARD_TOTAL=0 TEST_RUN='TestCommandsAdministrationCollStatsWithScale'
task test-integration-postgresql SHARD_INDEX=0 SHARD_TOTAL=0 TEST_RUN='TestCommandsAdministrationDataSize'
```

See also this issue mentioned in the code - https://github.com/FerretDB/FerretDB/pull/3522/files.

### Definition of Done

- fixes provided;
- integration/compatibility tests updated.

### PR DESCRIPTION
**Pull request by chilagrow**
# Description

Closes #2447.
Closes #3521.

This is analysis of the issue.
Keep https://github.com/FerretDB/FerretDB/issues/2447 open, capped collection related things can't be tested until https://github.com/FerretDB/FerretDB/issues/3458 is done, `freeStorageSize` needs to be implemented.

## Readiness checklist

- [x] I added/updated unit tests (and they pass).
- [x] I added/updated integration/compatibility tests (and they pass).
- [x] I added/updated comments and checked rendering.
- [x] I made spot refactorings.
- [ ] I updated user documentation.
- [ ] I ran `task all`, and it passed.
- [x] I ensured that PR title is good enough for the changelog.
- [x] (for maintainers only) I set Reviewers ([`@FerretDB/core`](https://github.com/orgs/FerretDB/teams/core)), Milestone (`Next`), Labels, Project and project's Sprint fields.
- [x] I marked all done items in this checklist.

### CODE REVIEW

File: internal/handlers/common/aggregations/stages/collstats.go [hunk 1]
```diff
@@ -70,14 +70,18 @@ func newCollStats(stage *types.Document) (aggregations.Stage, error) {
 
 		storageStatsFields := must.NotFail(fields.Get("storageStats")).(*types.Document)
 
+		cs.storageStats.scale = 1
```
Comment (chilagrow) [inline internal/handlers/common/aggregations/stages/collstats.go:73]: Adding test revealed that it was set zero.

Comment (chilagrow) [reply internal/handlers/common/aggregations/stages/collstats.go]: Never mind, it will be handled in https://github.com/FerretDB/FerretDB/pull/3565 PR to avoid conflict

File: integration/commands_administration_compat_test.go [hunk 2]
```diff
@@ -97,6 +100,67 @@ func TestCommandsAdministrationCompatCollStatsWithScale(t *testing.T) {
...
+		AddNonExistentCollection: true,
+	})
+
+	ctx, targetCollection, compatCollection := s.Ctx, s.TargetCollections[0], s.CompatCollections[0]
+
+	for name, tc := range map[string]struct { //nolint:vet // for readability
+		opts *options.CreateCollectionOptions
+	}{
+		"Size": {
+			opts: &options.CreateCollectionOptions{
```
Comment (AlekSi) [inline integration/commands_administration_compat_test.go:119]: `options.CreateCollection().SetCapped`

Please remove all `&options.` in this PR

File: internal/handlers/sqlite/msg_aggregate.go [hunk 3]
```diff
@@ -408,9 +408,9 @@ func processStagesStats(ctx context.Context, closer *iterator.MultiCloser, p *st
 				"count", collStats.CountObjects,
 				"avgObjSize", avgObjSize,
 				"storageSize", collStats.SizeCollection,
-				"freeStorageSize", int64(0), // TODO https://github.com/FerretDB/FerretDB/issues/2342
-				"capped", false, // TODO https://github.com/FerretDB/FerretDB/issues/2342
-				"wiredTiger", must.NotFail(types.NewDocument()), // TODO https://github.com/FerretDB/FerretDB/issues/2342
+				// TODO https://github.com/FerretDB/FerretDB/issues/2447
+				"freeStorageSize", int64(0),
+				"capped", collStats.CappedSize != int64(0),
```
Comment (AlekSi) [inline internal/handlers/sqlite/msg_aggregate.go:413]:
```suggestion
				"capped", collStats.CappedSize > 0,
```

That's what we use already

File: internal/handlers/sqlite/msg_collstats.go [hunk 4]
```diff
@@ -97,6 +97,11 @@ func (h *Handler) MsgCollStats(ctx context.Context, msg *wire.OpMsg) (*wire.OpMs
 		pairs = append(pairs, "avgObjSize", stats.SizeCollection/stats.CountObjects)
 	}
 
+	capped := stats.CappedSize != int64(0)
```
Comment (AlekSi) [inline internal/handlers/sqlite/msg_collstats.go:100]:
```suggestion
	capped := stats.CappedSize > 0
```

File: internal/backends/collection.go [hunk 5]
```diff
@@ -227,15 +227,16 @@ type CollectionStatsParams struct {
...
-// TODO https://github.com/FerretDB/FerretDB/issues/2447
 type CollectionStatsResult struct {
-	CountObjects   int64
-	CountIndexes   int64
-	SizeTotal      int64
-	SizeIndexes    int64
-	SizeCollection int64
+	CappedSize      int64
+	CappedDocuments int64
+	CountObjects    int64
```
Comment (AlekSi) [inline internal/backends/collection.go:235]: What is the difference between those two?

And let's finally rename "objects" to "documents"

File: internal/handlers/sqlite/msg_aggregate.go (hunk 3 above)
Comment (chilagrow) [reply internal/handlers/sqlite/msg_aggregate.go]: At the end, this was replaced by .Capped() of CollectionInfo

File: internal/backends/collection.go (hunk 5 above)
Comment (chilagrow) [reply internal/backends/collection.go]: renamed, also getting capped document info from ListCollections now

File: internal/handlers/sqlite/msg_collstats.go [hunk 6]
```diff
@@ -76,6 +76,25 @@ func (h *Handler) MsgCollStats(ctx context.Context, msg *wire.OpMsg) (*wire.OpMs
 		return nil, lazyerrors.Error(err)
 	}
 
+	list, err := db.ListCollections(ctx, new(backends.ListCollectionsParams))
```
Comment (rumyantseva) [inline internal/handlers/sqlite/msg_collstats.go:79]: So, here we call `ListCollection` only because we want to find the collection info of a particular collection.
I see just calling `db.Collection()` is not possible because it returns `Collection` and not `CollectionInfo`.

Did we consider an approach when we can perform something like `db.Collection(collection).Info()` to get `CollectionInfo` of a particular collection? Calling the whole list and iterating through it when we only need one collection doesn't look super reasonable.

File: internal/handlers/sqlite/msg_aggregate.go [hunk 7]
```diff
@@ -394,23 +396,42 @@ func processStagesStats(ctx context.Context, closer *iterator.MultiCloser, p *st
 		if err != nil {
 			return nil, lazyerrors.Error(err)
 		}
+
+		var list *backends.ListCollectionsResult
+
+		if list, err = p.db.ListCollections(ctx, new(backends.ListCollectionsParams)); err != nil {
```
Comment (rumyantseva) [inline internal/handlers/sqlite/msg_aggregate.go:402]: Same as for `msg_collstats`: maybe we don't need to query all the collections if we need only one.

**Review by rumyantseva — COMMENTED** [review]
Great job overall, left one suggestion.

File: internal/handlers/sqlite/msg_collstats.go (hunk 6 above)
Comment (AlekSi) [reply internal/handlers/sqlite/msg_collstats.go:79]: Why does that not look reasonable?

File: internal/handlers/sqlite/msg_aggregate.go (hunk 7 above)
Comment (AlekSi) [reply internal/handlers/sqlite/msg_aggregate.go:402]: `ListCollections` does not query anything

File: internal/backends/collection.go [hunk 8]
```diff
@@ -227,11 +227,9 @@ type CollectionStatsParams struct {
 
 // CollectionStatsResult represents the results of Collection.Stats method.
 //
-// CountObjects is an estimate of the number of documents.
-//
-// TODO https://github.com/FerretDB/FerretDB/issues/2447
+// CountDocuments is an estimate of the number of documents.
```
Comment (AlekSi) [inline internal/backends/collection.go:230]: That comment was only needed because it wasn't clear what are "objects". I think it is already clear now

File: internal/backends/collection.go [hunk 9]
```diff
@@ -227,11 +227,9 @@ type CollectionStatsParams struct {
...
 type CollectionStatsResult struct {
-	CountObjects   int64
+	CountDocuments int64
 	CountIndexes   int64
```
Comment (AlekSi) [inline internal/backends/collection.go:233]: We probably should remove that

File: internal/handlers/sqlite/msg_collstats.go (hunk 6 above)
Comment (rumyantseva) [reply internal/handlers/sqlite/msg_collstats.go:79]: Because inside `r.CollectionList` we iterate through all the collections in the DB (`r.colls`) to make a slice with response data, and then we iterate through this resulting slice once again (here) to find one particular collection by its name. Instead of that, we can just get the exact collection from the map  `r.colls` by its name.

Comment (AlekSi) [reply internal/handlers/sqlite/msg_collstats.go:79]: And add a method for every backend to implement that does not add anything new.

I don't think that iterating over memory is worse than that.

File: internal/backends/collection.go (hunk 9 above)
Comment (chilagrow) [reply internal/backends/collection.go]: Removed it also from DBStats.

File: integration/commands_administration_compat_test.go [hunk 10]
```diff
@@ -97,6 +100,78 @@ func TestCommandsAdministrationCompatCollStatsWithScale(t *testing.T) {
...
+		Providers:                []shareddata.Provider{},
+		AddNonExistentCollection: true,
+	})
+
+	ctx, targetCollection, compatCollection := s.Ctx, s.TargetCollections[0], s.CompatCollections[0]
+
+	for name, tc := range map[string]struct { //nolint:vet // for readability
+		capped       *bool
+		sizeInBytes  *int64
+		maxDocuments *int64
```
[the remark this answers is not in the document]
Comment (chilagrow) [reply integration/commands_administration_compat_test.go]: Simplified it :+1:

File: integration/commands_administration_compat_test.go [hunk 11]
```diff
@@ -97,6 +100,78 @@ func TestCommandsAdministrationCompatCollStatsWithScale(t *testing.T) {
...
+			maxDocuments: pointer.ToInt64(10),
+		},
+	} {
+		name, tc := name, tc
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+
+			cName := testutil.CollectionName(t) + name
+			opts := options.CreateCollection()
+			if tc.capped != nil {
```
[the remark this answers is not in the document]
Comment (chilagrow) [reply integration/commands_administration_compat_test.go]: Also simplified this too :+1:

File: internal/backends/collection.go (hunk 8 above)
Comment (chilagrow) [reply internal/backends/collection.go]: Removed comment

**Review by noisersup — APPROVED** [review]
LGTM

**Review by AlekSi — APPROVED** [review]

File: internal/handlers/sqlite/msg_dbstats.go [hunk 12]
```diff
@@ -61,6 +61,36 @@ func (h *Handler) MsgDBStats(ctx context.Context, msg *wire.OpMsg) (*wire.OpMsg,
...
 	}
 
+	list, err := db.ListCollections(ctx, new(backends.ListCollectionsParams))
+	if err != nil {
+		return nil, lazyerrors.Error(err)
+	}
+
+	var nIndexes int64
+
+	for _, cInfo := range list.Collections {
```
Comment (AlekSi) [inline internal/handlers/sqlite/msg_dbstats.go:71]:
```diff
- for _, cInfo := range list.Collections
+ for _, cInfo = range list.Collections
```

I will do that in my PR

File: internal/handlers/sqlite/msg_aggregate.go [hunk 13]
```diff
@@ -392,24 +395,57 @@ func processStagesStats(ctx context.Context, closer *iterator.MultiCloser, p *st
...
+
+		var cList *backends.ListCollectionsResult
+
+		if cList, err = p.db.ListCollections(ctx, new(backends.ListCollectionsParams)); err != nil {
+			return nil, lazyerrors.Error(err)
+		}
+
+		var found bool
+
+		for _, cInfo := range cList.Collections {
```
Comment (AlekSi) [inline internal/handlers/sqlite/msg_aggregate.go:407]:
```diff
- for _, cInfo := range list.Collections
+ for _, cInfo = range list.Collections
```

But then `cInfo` is not used?..
